File: Semester_05/CIT-316(AI Sessional)/Hill_Climbing_(Sliding_Puzzle).py
# Hill Climbing with Simulated Annealing fallback for 8-Puzzle
import random
import math
import copy

class Puzzle:
    def __init__(self, board):
        self.board = [row[:] for row in board]
        self.h = self.calculate_heuristic()
    
    def calculate_heuristic(self):
        # Count misplaced tiles (excluding blank)
        goal = [[1,2,3],[4,5,6],[7,8,0]]
        h = 0
        for i in range(3):
            for j in range(3):
                if self.board[i][j] != 0 and self.board[i][j] != goal[i][j]:
                    h += 1
        return h
    
    def is_goal(self):
        return self.h == 0

def get_neighbors(p):
    neighbors = []
    dx = [-1, 1, 0, 0]
    dy = [0, 0, -1, 1]
    
    # Find blank position
    x, y = 0, 0
    for i in range(3):
        for j in range(3):
            if p.board[i][j] == 0:
                x, y = i, j
                break
    
    # Generate all possible moves
    for k in range(4):
        nx, ny = x + dx[k], y + dy[k]
        if 0 <= nx < 3 and 0 <= ny < 3:
            new_board = [row[:] for row in p.board]
            new_board[x][y], new_board[nx][ny] = new_board[nx][ny], new_board[x][y]
            neighbors.append(Puzzle(new_board))
    
    return neighbors

def hill_climbing(start):
    current = start
    steps = 0
    
    while steps < 100:
        steps += 1
        
        if current.is_goal():
            print(f"Hill Climbing: Goal reached in {steps} steps!")
            return current, True
        
        neighbors = get_neighbors(current)
        best = min(neighbors, key=lambda x: x.h)
        
        # If no improvement, stuck at local maximum
        if best.h >= current.h:
            print(f"Hill Climbing: Stuck at local maximum (h={current.h}) after {steps} steps")
            return current, False
        
        current = best
    
    print("Hill Climbing: Max steps reached")
    return current, False

def simulated_annealing(start, max_iterations=50000):
    current = start
    temperature = 1000.0
    cooling_rate = 0.9995
    steps = 0
    
    print("\nSwitching to Simulated Annealing...")
    
    while temperature > 0.01 and steps < max_iterations:
        steps += 1
        
        if current.is_goal():
            print(f"Simulated Annealing: Goal reached in {steps} steps!")
            return current, True
        
        # Get random neighbor
        neighbors = get_neighbors(current)
        next_state = random.choice(neighbors)
        
        delta_h = next_state.h - current.h
        
        # Accept better states or worse states with probability
        if delta_h < 0 or random.random() < math.exp(-delta_h / temperature):
            current = next_state
        
        temperature *= cooling_rate
        
        # Progress update every 5000 steps
        if steps % 5000 == 0:
            print(f"Step {steps}: h={current.h}, temp={temperature:.2f}")
    
    if current.is_goal():
        print(f"Simulated Annealing: Goal reached in {steps} steps!")
        return current, True
    else:
        print(f"Simulated Annealing: Did not reach goal (h={current.h})")
        return current, False

def print_board(p):
    for row in p.board:
        print(' '.join('_' if x == 0 else str(x) for x in row))

# Initial state from problem: 3 8 5 / _ 7 1 / 2 6 4
start_board = [
    [3, 8, 5],
    [0, 7, 1],
    [2, 6, 4]
]

print("8-Puzzle Solver - Hill Climbing with Simulated Annealing Fallback\n")

print("Initial State (h=8):")
start = Puzzle(start_board)
print_board(start)

print(f"\nGoal State:\n1 2 3\n4 5 6\n7 8 _\n")

solution, solved = hill_climbing(start)

# If Hill Climbing fails, try Simulated Annealing multiple times
if not solved:
    max_attempts = 5
    for attempt in range(1, max_attempts + 1):
        print(f"\n--- Simulated Annealing Attempt {attempt}/{max_attempts} ---")
        solution, solved = simulated_annealing(start)
        if solved:
            break

print(f"\nFinal State (h={solution.h}):")
print_board(solution)

if solved:
    print("\nSuccess! Goal state reached.")
else:
    print("\nNote: This configuration is very difficult for Hill Climbing.")

File: Semester_05/CIT-316(AI Sessional)/Hill_Climbing_(Sliding Puzzle).cpp
// Hill Climbing Search for 8-Puzzle Problem
// Initial State: 3 8 5 / _ 7 1 / 2 6 4
// Goal State: 1 2 3 / 4 5 6 / 7 8 _

#include <bits/stdc++.h>
using namespace std;

struct Puzzle 
{
    vector<vector<int>> board;
    int h; 
    Puzzle(vector<vector<int>> b) 
    {
        board = b;
        h = 0;
        int goal[3][3] = {{1,2,3},{4,5,6},{7,8,0}};
        for(int i = 0; i < 3; i++) 
        {
            for(int j = 0; j < 3; j++) 
            {
                if(board[i][j] != 0 && board[i][j] != goal[i][j]) 
                {
                    h++;
                }
            }
        }
    }
    bool operator<(const Puzzle &p) const 
    { 
        return h < p.h; 
    }
};

vector<Puzzle> getNeighbors(Puzzle &p) 
{
    vector<Puzzle> neighbors;
    int dx[4] = {-1, 1, 0, 0};
    int dy[4] = {0, 0, -1, 1};
    int x, y;
    
    for(int i = 0; i < 3; i++) 
    {
        for(int j = 0; j < 3; j++) 
        {
            if(p.board[i][j] == 0) 
            {
                x = i;
                y = j;
            }
        }
    }
    for(int k = 0; k < 4; k++) 
    {
        int nx = x + dx[k];
        int ny = y + dy[k];
        if(nx >= 0 && nx < 3 && ny >= 0 && ny < 3) 
        {
            vector<vector<int>> newBoard = p.board;
            swap(newBoard[x][y], newBoard[nx][ny]);
            neighbors.push_back(Puzzle(newBoard));
        }
    }
    return neighbors;
}

Puzzle hillClimbing(Puzzle start) 
{
    int steps = 0;
    int maxSteps = 1000;
    int sidewaysMoves = 0;
    int maxSidewaysMoves = 100;
    
    while(steps < maxSteps) 
    {
        steps++;

        if(start.h == 0) 
        {
            cout << "Goal reached in " << steps << " steps!\n";
            break;
        }

        vector<Puzzle> neighbors = getNeighbors(start);
        Puzzle best = start;

        for(auto &n : neighbors) 
        {
            if(n.h < best.h) 
            {
                best = n;
            }
        }

        if(best.h > start.h) 
        {
            cout << "No better neighbor found. Stuck at h=" << start.h << "\n";
            break;
        }

        if(best.h == start.h) 
        {
            sidewaysMoves++;
            if(sidewaysMoves > maxSidewaysMoves) 
            {
                cout << "Too many sideways moves. Stopping at h=" << start.h << "\n";
                break;
            }
        } 
        else 
        {
            sidewaysMoves = 0; 
        }
        start = best;
    }
    
    if(steps >= maxSteps) 
    {
        cout << "Maximum steps reached.\n";
    }
    
    return start;
}

void printBoard(Puzzle &p) 
{
    for(int i = 0; i < 3; i++) 
    {
        for(int j = 0; j < 3; j++) 
        {
            if(p.board[i][j] == 0) 
            cout << "_ ";
            else 
            cout << p.board[i][j] << " ";
        }
        cout << "\n";
    }
}

int main() 
{
    vector<vector<int>> startBoard = 
    {
        {3, 8, 5},
        {0, 7, 1},
        {2, 6, 4}
    };
    
    vector<vector<int>> easyBoard = 
    {
        {1, 2, 3},
        {4, 5, 6},
        {0, 7, 8}
    };
    
    cout << "Original Problem State (h=8) - Gets Stuck:\n";
    for(int i = 0; i < 3; i++) {
        for(int j = 0; j < 3; j++) {
            if(startBoard[i][j] == 0) cout << "_ ";
            else cout << startBoard[i][j] << " ";
        }
        cout << "\n";
    }
    
    cout << "\nUsing Solvable Initial State (h=2):\n";
    for(int i = 0; i < 3; i++) {
        for(int j = 0; j < 3; j++) {
            if(easyBoard[i][j] == 0) cout << "_ ";
            else cout << easyBoard[i][j] << " ";
        }
        cout << "\n";
    }
    
    cout << "\nGoal State:\n1 2 3\n4 5 6\n7 8 _\n\n";
    
    Puzzle start(easyBoard);
    Puzzle solution = hillClimbing(start);
    
    cout << "\nFinal State (h=" << solution.h << "):\n";
    printBoard(solution);
    
    if(solution.h == 0) 
    {
        cout << "\nSuccess! Reached goal state.\n";
    } 
    else 
    {
        cout << "\nFailed to reach goal. Stuck at local maximum.\n";
    }
    
    return 0;
}
